### ISSUE TITLE
Really supporting 486 instructions

### ISSUE THREAD

**Issue by roytam1** [captainys/TOWNSEMU#140]
Tsugaru's so-called 486 CPU doesn't support `lock xadd [eax], edx` and throws SIGILL to guest OS.
![image](https://github.com/user-attachments/assets/0b22b246-9c10-404c-9cec-d0ff739ded97)

**Comment by roytam1** [comment]
trying to hack around ATL and MFC to not using `intrinsic(_InterlockedIncrement)` and `intrinsic(_InterlockedDecrement)` and let it use winapi call for them, instead of throwing `SIGILL` VM aborts on opcode `cpuid`:
![image](https://github.com/user-attachments/assets/04f26b77-de66-4112-8573-5ba0dbb19452)

test binary: [CrystalMarkRetro2_0_0Beta2_3_9x.zip](https://github.com/user-attachments/files/18181429/CrystalMarkRetro2_0_0Beta2_3_9x.zip)

**Comment by captainys** [comment]
Thank you for finding an unsupported instruction!  Are you able to build Tsugaru on your end?  I've just pushed the source that I hope supports XADD.  I've been adding support for instructions as I encounter.  Your contribution for finding and testing unsupported instructions is very hepful.
Thanks!

**Comment by roytam1** [comment]
I didn't tried to build any (I don't know if my environment is capable to so do, Win7 x64 + VS 2017 here).
I wonder if Github Actions can provide Artifacts? AFAIK some projects like https://github.com/reactos/reactos/actions/runs/12385435979 can provide Artifacts for each run.

**Comment by binarymaster** [comment]
@roytam1 Tsugaru repository uses GitHub Actions as well, just doesn't upload artifacts, see [build-windows](https://github.com/captainys/TOWNSEMU/blob/master/.github/workflows/continuous.yml).

**Comment by captainys** [comment]
I don't think I used anything that is not supported by VS 2017, but I haven't tried, so cannot say for sure.

Logging undefined instructions and shoot INT 6 can be a good idea, or I can make it an option to abort or to shoot INT 6.

I thought there is a limit on the total size of artifacts for free account.  I may be wrong though.

**Comment by roytam1** [comment]
> Logging undefined instructions and shoot INT 6 can be a good idea, or I can make it an option to abort or to shoot INT 6.

yeah, please do this. either option is fine to me (for the switchable option, I hope shoot INT 6 will be the default option)

**Comment by bcc2528** [comment]
XADD命令実装のコードを試してみましたが、CrystalMarkRetroは「SHWLAPI.DLLが見つかりません」エラーで止まります。
どうやらCrystalMarkRetroはInternet Explorer 4.0が必要なのに対して、FM TOWNS版Windows 95はInternet Explorer 2.0しかプリインストールされていないのが原因のようです。今からIE4.0の入手できる
場所は調べておらず、第一TOWNS版95にインストールできるか未知数ではありますが。

CrystalMarkRetro requires Internet Explorer 4.0, but the version of Internet Explorer in Windows 95 for FM TOWNS is 2.0.
![CrystalDiskMark](https://github.com/user-attachments/assets/3743b70e-94ce-457d-939c-b36a55f3068d)

**Comment by roytam1** [comment]
> CrystalMarkRetro requires Internet Explorer 4.0, but the version of Internet Explorer in Windows 95 for FM TOWNS is 2.0.

where did you get the CMR archive? the one in https://github.com/captainys/TOWNSEMU/issues/140#issuecomment-2551245643 should not depend on shlwapi.dll

**Comment by bcc2528** [comment]
公式サイトからではなく#140のファイルを使ったら実行できましたが、やはりCPUID命令が必要なようですね。486では後期ロットで実装されたようなのでCPU換装をしていないMA/MXでは動かないかもしれません。
そのままCPUID実装してもいいのかもしれませんが、フラグレジスタの21ビット目が変更できるかどうかで命令の有無判定するようなのでそこだけ弄れば動いたりして。

I was able to run it using your file instead of from the official website, but it seems that the CPUID instruction is still required.

**Comment by roytam1** [comment]
> I was able to run it using your file instead of from the official website, but it seems that the CPUID instruction is still required.

Package in https://github.com/captainys/TOWNSEMU/issues/140#issuecomment-2551245643 are tested by an owner of 386-class PC-98 which works: https://x.com/drachen6jp/status/1869410721936888051

so the problem is not in missing `cpuid` but the handling of illegal instruction.

**Comment by roytam1** [comment]
and another example is CPU-Z vintage edition: https://www.cpuid.com/softwares/cpu-z.html
which tries to execute VIA's `JMPAI` (i.e. `0f 3f`) to probe CPU capabilities.
![image](https://github.com/user-attachments/assets/8b6102f4-2ce8-41a4-9d60-1a3e90a02e40)

````
Probably call from Windows 95 CD-ROM driver.
VM Aborted!
Device:486DX
Reason:Clocks-Passed is not set.  Opcode=0F3FH
Towns TIME (Nano-Seconds): 62533842425
CS:EIP=0137:0054EBDC  LINEAR:0054EBDC  EFLAGS=00000203  CPL=03
EAX=00000001  EBX=00000000  ECX=00000000  EDX=00000000
ESI=00BF1890  EDI=00000001  EBP=009BF470  ESP=009BF43C
CS=0137(LIN:00000000)  DS=013F(LIN:00000000)  ES=013F(LIN:00000000)
FS=22EF(LIN:8157886C)  GS=0000(LIN:00000000)  SS=013F(LIN:00000000)
CR0=8000002D  CR1=00000000  CR2=8157E1F8  CR3=0030D000
CF1  PF0  AF0  ZF0  SF0  TF0  IF1  DF0  OF0  IOPL00  NT0  RF0  VM0  AC0
Default Operand Size=32  Default Address Size=32  Stack Address Size=32
SS+00000000:00 00 00 00 01 00 00 00 90 18 BF 00 00 00 00 00
SS+00000010:00 00 00 00 3F F0 54 00 00 F4 9B 00 40 F4 9B 00
0137:0054EBDC 0F3F
>
````

So CPU should not be abort by default.

**Comment by bcc2528** [comment]
![CrystalMarkRetro](https://github.com/user-attachments/assets/ca6e2f30-1924-4d82-a496-1845e0b5c022)
CPUID命令実装でCrystalMarkRetroは動きました。
486でのCPUID命令実装はおそらくDX4からと思われますが、95年1月付のデータシート384ページに情報が載っています。
https://bitsavers.org/components/intel/80486/241731-002_Intel486_Microprocessors_and_Related_Products_Jan95.pdf

CPU-Zは0f 3f(VIA CPU専用のデバッグ用命令で本来なら非公開)で止まりますが、VIA CPUだと誤判定したとしてもこの命令が使うとも思えないのでTOWNS機で動かした際の特有の不具合の可能性もあるのかも。PC-98でCPU-Zを動かすとキーボードが受け付けなくなるというバグもあるようだし。
